editor/src/components/canvas/dom-walker.ts

```diff
@@ -94,24 +100,136 @@ function isScene(node: Node): node is HTMLElement {
+            invalidatedSceneIDsRef.current != null &&
+            invalidatedSceneIDsRef.current.indexOf(sceneID) < 0
+          ) {
+            invalidatedSceneIDsRef.current.push(sceneID)
+          }
+        }
+      })
+    } else {
+      return null
+    }
+    /* eslint-disable-next-line react-hooks/exhaustive-deps */
+  }, []) // the dependencies are empty because this should only evaluate once
+  React.useEffect(() => {
+    return function cleanup() {
+      if (resizeObserver != null) {
+        resizeObserver.disconnect()
+      }
+    }
+  }, [resizeObserver])
+  return resizeObserver
+}
+
+function useMutationObserver(invalidatedSceneIDsRef: React.MutableRefObject<Array<string>>) {
+  const mutationObserver = React.useMemo(() => {
+    if (ObserversAvailable) {
+      return new (window as any).MutationObserver((mutations: MutationRecord[]) => {
+        for (let mutation of mutations) {
+          if (
+            mutation.attributeName === 'style' ||
+            mutation.addedNodes.length > 0 ||
+            mutation.removedNodes.length > 0
+          ) {
+            if (mutation.target instanceof HTMLElement) {
+              const sceneID = findParentScene(mutation.target)
+              if (
+                sceneID != null &&
+                invalidatedSceneIDsRef.current != null &&
+                invalidatedSceneIDsRef.current.indexOf(sceneID) < 0
+              ) {
+                invalidatedSceneIDsRef.current.push(sceneID)
+              }
+            }
+          }
+        }
+      })
+    } else {
+      return null
+    }
+    /* eslint-disable-next-line react-hooks/exhaustive-deps */
+  }, []) // the dependencies are empty because this should only evaluate once
+  React.useEffect(() => {
+    return function cleanup() {
+      if (mutationObserver != null) {
+        mutationObserver.disconnect()
+      }
+    }
+  }, [mutationObserver])
+  return mutationObserver
+}
+
 export function useDomWalker(props: CanvasContainerProps): React.Ref<HTMLDivElement> {
   const containerRef = React.useRef<HTMLDivElement>(null)
+  const rootMetadataInStateRef = useRefEditorState((store) => store.editor.domMetadataKILLME)
+  const invalidatedSceneIDsRef = React.useRef<Array<string>>([])
+  const initCompleteRef = React.useRef<boolean>(false)
   const selectedViews = useEditorState(
     (store) => store.editor.selectedViews,
     'useDomWalker selectedViews',
   )
+  const resizeObserver = useResizeObserver(invalidatedSceneIDsRef)
+  const mutationObserver = useMutationObserver(invalidatedSceneIDsRef)
 
   React.useLayoutEffect(() => {
     if (containerRef.current != null) {
+      if (LogDomWalkerPerformance) {
+        performance.mark('DOM_WALKER_START')
+      }
       // Get some base values relating to the div this component creates.
       const refOfContainer = containerRef.current
+      if (ObserversAvailable) {
+        Array.from(document.querySelectorAll('#canvas-container *')).map((elem) => {
```

Comment:
Please can you make `'canvas-container'` a `const` that is used here and on the actual canvas container

Reply:
fixed